### PR TITLE
perf(dsv4 hc_pre): split-K decode path (~1.8x), keep fused prefill

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary
- Decode (`T = B*S = 8`) ran the fused single-spmd `hc_pre` on **one core** (1 of 24 AIC, 1 of 48 AIV) — one token-tile is one `pl.spmd` block. `hc_pre` now **dispatches on `T` at runtime**: `_hc_pre_decode` (split-K + per-axis fan-out) for small T, `_hc_pre_prefill` (the fused single-task, #533) for large T. It inlines into each decode/prefill attention kernel, so each context compiles only its branch.
- `_hc_pre_decode` mirrors `hc_head`'s pure-AIC split-K: cast `x -> x_fp32`, split the `K=16384` projection into `LINEAR_OK` slices that atomic-add FP32 partials (1 cube task -> 8), fan the cast over K and `mix_x` over D, and keep the 20-iter Sinkhorn as its own serial scope (a latency floor more cores can't shorten).
- Prefill keeps the fused task: its token-tiles already fill the chip, so the decode fan-out would only add AICPU dispatch overhead (it ballooned to ~288 task-instances and regressed prefill ~11%).
- Device a2a3 (910B), golden-validated both modes (`x_mixed`/`post`/`comb` `ratio_allclose`), best-of-N: **decode 125us -> ~70us (~1.8x)** (matmul 40->7us, 42us BF16 pad removed); **prefill 147us -> ~143us (flat, no regression)**.

## Related Issues
None.

### CODE REVIEW

File: models/deepseek/v4/hc_pre.py [hunk 1]
```diff
@@ -41,41 +63,246 @@
...
+            else:
+                acc = pl.matmul_acc(acc, x_linear_chunk, w_chunk, b_trans=True)
+        mixes_raw = pl.assemble(mixes_raw, acc, [t0, 0], atomic=pl.AtomicType.Add)
+
+    # Scale the raw projection by inv_rms, then sigmoid -> pre (kept for mix_x via
+    # pre_val_store) and post (written straight out); comb logits bridge to the
+    # sinkhorn scope. inv_rms is folded in per column-group (cheap on [T, <=16]).
+    pre_val_store = pl.create_tensor([T_MAX, HC_PAD], dtype=pl.FP32)
+    post_pad_store = pl.create_tensor([T_MAX, HC_PAD], dtype=pl.FP32)
+    comb_logits = pl.create_tensor([T_MAX, HC_MULT * HC_MULT], dtype=pl.FP32)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply models/deepseek/v4/hc_pre.py]: Fixed in 1fa8232 — widened `comb_logits` to `MIX_PAD` (32, same as `mixes_raw`) so every comb-group load descriptor (`k*HC_MULT` offset, `HC_PAD` wide) stays in-bounds; group 3's [12:20] span no longer exceeds the tensor. Decode + prefill golden re-validated, ruff clean.

File: models/deepseek/v4/hc_pre.py [hunk 2]
```diff
@@ -41,41 +65,265 @@
...
+
+        row_sum_tmp_iter = pl.create_tile([COMB_T_TILE, 1], dtype=pl.FP32, target_memory=pl.MemorySpace.Vec)
+        col_sum = pl.add(pl.add(row0_eff, row1_eff), pl.add(row2_eff, row3_eff))
+        col_sum = pl.add(col_sum, HC_EPS)
+        row0_cur = pl.div(row0_eff, col_sum)
+        row1_cur = pl.div(row1_eff, col_sum)
+        row2_cur = pl.div(row2_eff, col_sum)
+        row3_cur = pl.div(row3_eff, col_sum)
+
+        for sk_it in pl.pipeline(HC_SINKHORN_ITER - 1, stage=2):
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply models/deepseek/v4/hc_pre.py]: Fixed in 0128859 — renamed to `_sk_it` (B007).

File: models/deepseek/v4/hc_pre.py [hunk 3]
```diff
@@ -41,41 +68,261 @@
...
+        acc = pl.create_tensor([LINEAR_T_TILE, MIX_PAD], dtype=pl.FP32)
+        for kb in pl.pipeline(0, LINEAR_CHUNKS_PER_SPLIT, stage=2):
+            k0 = k_base + kb * LINEAR_K_CHUNK
+            x_linear_chunk = x_fp32[t0:t0 + LINEAR_T_TILE, k0:k0 + LINEAR_K_CHUNK]
+            w_chunk = pl.slice(hc_fn, [MIX_PAD, LINEAR_K_CHUNK], [0, k0], valid_shape=[MIX_HC, LINEAR_K_CHUNK])
+            if kb == 0:
+                acc = pl.matmul(x_linear_chunk, w_chunk, b_trans=True, out_dtype=pl.FP32)
+            else:
+                acc = pl.matmul_acc(acc, x_linear_chunk, w_chunk, b_trans=True)
+        mixes_raw = pl.assemble(mixes_raw, acc, [t0, 0], atomic=pl.AtomicType.Add)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply models/deepseek/v4/hc_pre.py:183]: Deliberate trade-off, not an oversight. The split-K decode accumulation uses `assemble(atomic=Add)` for the ~10us decode win (125->70us); `hc_head` uses the same primitive. a2a3sim/a5sim don't model the atomic accumulate, so those two sim jobs fail by design and are skipped for this PR — `a2a3` (device) and `unit-tests` are green and the device path is golden-validated (decode B4S2, prefill B1S128). A `# ci: no-sim` tag wouldn't be a clean fix: the decode branch (atomic) is inlined into the decode-side dependents too (decode_attention_*, decode_layer, moe), so tagging to skip would drop sim coverage across all of them, not just hc_pre.py.

**Comment by Hzfengsy** [comment]
Close as 653 applies the same optimization
